src/logger.hpp

```diff
@@ -0,0 +1,174 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2020 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+#pragma once
+#include <iostream>
+#include <memory>
+#include <queue>
+
+#include "sync/sync_manager.hpp"  // SyncLoggerFactory
+#include "util/scheduler.hpp"     // realm::util::Scheduler
+
+#if REALM_ANDROID
+#include <android/log.h>
+#endif
+
+namespace realm {
+namespace common {
+namespace logger {
+
+using LoggerLevel = realm::util::Logger::Level;
+
+// TODO we are coupling core with JS Land here, change to string.
+using Entry = std::pair<LoggerLevel, std::string>;
+using Delegated = std::function<void(int, std::string)>;
+
+/*
+ * The idea here is to one day implement a functionality to delegate the logs
+ * to the mobile OS, for that we got two specialized behaviour that we can
+ * inject at compile time or at runtime.
+ *
+ */
+
+#if REALM_ANDROID
+class AndroidLogger {
+    std::map<LoggerLevel, android_LogPriority> map_android_log_level{
+        {LoggerLevel::all, ANDROID_LOG_VERBOSE},
+        {LoggerLevel::info, ANDROID_LOG_INFO},
+        {LoggerLevel::trace, ANDROID_LOG_DEFAULT},
+        {LoggerLevel::debug, ANDROID_LOG_DEBUG},
+        {LoggerLevel::detail, ANDROID_LOG_VERBOSE},
+        {LoggerLevel::warn, ANDROID_LOG_WARN},
+        {LoggerLevel::error, ANDROID_LOG_ERROR},
+        {LoggerLevel::fatal, ANDROID_LOG_FATAL},
+        {LoggerLevel::off, ANDROID_LOG_SILENT},
+    };
+
+    void print(Entry& entry) {
+        auto android_log_level = map_android_log_level[entry.first];
+        __android_log_print(android_log_level, "realm", "%s",
+                            entry.second.c_str());
+    }
+};
+#endif
+
+#ifdef __APPLE__
+class IOSLogger {
+    void print() {}
+};
+#endif
+
+class SyncLoggerDelegator : public realm::util::RootLogger {
+   public:
+    void delegate(Delegated& delegate) {
+        m_scheduler->set_notify_callback([this, delegate] {
+            while (!m_log_queue.empty()) {
+                /*
+                 *  Extracting and delegating log entries.
+                 */
+                std::queue<Entry> popped;
```

Comment:
the swap should be done outside the while loop. then the iteration should be over the `popped` queue.